feat(expression router) Implement translator of ingresses

randmonkey opened the pull request:
<!-- Thanks for sending a pull request! Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://github.com/kubernetes/community/blob/master/contributors/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://github.com/kubernetes/community/blob/master/contributors/guide/release-notes.md and ensure your changes are being reflected in CHANGELOG.md for the next upcoming release
-->

**What this PR does / why we need it**:

- add feature gate `ExpressionRoutes` and flags in related structures (`KongClient`,`Parser`,`KongState`,`Route`) to process expression based routes in a different way
- Implement translation of ingresses
- Add integration tests for expression based router

<!-- Please describe why this particular PR is necessary or why you see it as a nice addition -->

**Which issue this PR fixes**:

fixes #3750 

<!--
Here you can add any links to issues that this PR is relevant for.
You can use Github keywords (like: closes, fixes or resolves) to auto-resolve
the linked issue(s) when this PR gets merged.

For example: fixes #<issue number>
-->

**Special notes for your reviewer**:

<!-- Here you can add any open questions or notes that you might have for reviewers -->

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [ ] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

czeslavo commented on internal/dataplane/parser/translators/ingress_atc.go:
```diff
@@ -0,0 +1,243 @@
+	validMethods = regexp.MustCompile(`\A[A-Z]+$`)
+
+	// hostnames are complicated. shamelessly cribbed from https://stackoverflow.com/a/18494710
+	// TODO if the Kong core adds support for wildcard SNI route match criteria, this should change.
```
Ditto

randmonkey replied:
I think we should put this to GH issue when kong support this.

czeslavo commented on internal/dataplane/parser/translators/ingress_atc.go:
```diff
@@ -0,0 +1,243 @@
+}
+
+// hostMatcherFromIngressHosts translates hosts in IngressHost format to ATC matcher that matches any of them.
+func hostMatcherFromIngressHosts(hosts []string) atc.Matcher {
```
IMHO all of the `*MatcherFrom*` funcs deserve a separate unit test looking at their quite complex logic. Would you mind adding those?

randmonkey replied:
Added unit tests.

czeslavo commented on internal/manager/run.go:
```diff
@@ -158,6 +158,13 @@ func Run(ctx context.Context, c *Config, diagnostic util.ConfigDumpDiagnostic, d
 	}
 
+	if enabled, ok := featureGates[featuregates.ExpressionRoutesFeature]; ok && enabled {
+		if routerFlavor == "expressions" {
+			dataplaneClient.EnableExpressionRoutes()
+			setupLog.Info("expression routes mode has been enabled")
+		}
```
```suggestion
		if routerFlavor == "expressions" {
			dataplaneClient.EnableExpressionRoutes()
			setupLog.Info("expression routes mode has been enabled")
		} else {
		        setupLog.Info(fmt.Sprintf("ExpressionRoutes feature gate enabled, but Gateway run with %q router flavor, using this instead", routerFlavor))
		}
```

randmonkey replied:
done

czeslavo commented on internal/dataplane/kongstate/route.go:
```diff
@@ -255,17 +263,25 @@ func (r *Route) override(log logrus.FieldLogger, kongIngress *configurationv1.Ko
-			r.StripPath = nil
-			break
+
+	if !r.ExpressionRoutes {
+		for _, val := range r.Protocols {
+			if *val == "grpc" || *val == "grpcs" {
+				// grpc(s) doesn't accept strip_path
+				r.StripPath = nil
+				break
+			}
 		}
 	}
```
Could we move this logic into `normalizeProtocols`? That would save us at least one `if !r.ExpressionRoutes`. :)

randmonkey replied:
done

czeslavo commented on internal/dataplane/deckgen/generate.go:
```diff
@@ -48,7 +48,7 @@ func ToDeckContent(
 		for _, r := range s.Routes {
 			route := file.FRoute{Route: r.Route}
-			fillRoute(&route.Route)
+			fillRoute(&route.Route, k8sState.ExpressionRoutes)
```
I think that instead of putting the `ExpressionRoutes` feature flag into `KongState`, we could create some struct (e.g. `DeckContentGeneratingParams`) that would include it, as well as `schemas`, `selectorTags`, and `formatVersion` that are all passed to `ToDeckContent`. This way we'd keep the separation of the configuration that was generated by the Parser from the parameters that steer how the deck file is generated out of this config. WDYT?

randmonkey replied:
done

czeslavo reviewed (changes requested):
That's a lot of good work! 👍 

My comments mainly gravitate around my thought that we should be careful when adding such a volume of brand new code by making sure that all of the pieces we add are properly unit tested + we have at least one integration test covering it works with Kong as expected in a sunny day scenario.

randmonkey commented on internal/dataplane/parser/translators/ingress_test.go:
```diff
@@ -1259,7 +1259,7 @@ func TestTranslateIngress(t *testing.T) {
 				}
 			})
-			diff := cmp.Diff(tt.expected, TranslateIngress(tt.ingress, tt.addRegexPrefix), checkOnlyObjectMeta)
+			diff := cmp.Diff(tt.expected, TranslateIngress(tt.ingress, tt.addRegexPrefix, false), checkOnlyObjectMeta)
```
I will include them in the unit tests for translating ingresses into expression routes.

randmonkey replied:
Added unit tests.

randmonkey commented on internal/dataplane/kong_client.go:
```diff
@@ -63,6 +63,10 @@ type KongClient struct {
 	// the newer logic which combines them.
 	enableCombinedServiceRoutes bool
 
+	// enableExpressionRoutes indacates whether to the data-plane client will
```
done.

randmonkey commented on internal/manager/utils/kongconfig/root.go:
```diff
@@ -19,41 +19,56 @@ import (
+// - router flavor
 // - kong version.
-func ValidateRoots(roots []Root, skipCACerts bool) (string, kong.Version, error) {
+func ValidateRoots(roots []Root, skipCACerts bool) (string, string, kong.Version, error) {
```
changed this to a struct `KongStartUpOptions` including all kong startup config that affects KIC.

randmonkey commented on internal/manager/utils/kongconfig/root.go:
```diff
@@ -64,6 +79,22 @@ func DBModeFromRoot(r Root) (string, error) {
+	routerFlavor, ok := rootConfig["router_flavor"].(string)
+	if !ok {
+		return "", fmt.Errorf(
+			"invalid router_flavor configuration, expected a string got %t",
```
done

randmonkey commented on internal/dataplane/parser/translators/ingress_atc.go:
```diff
@@ -0,0 +1,243 @@
+
+var (
+	headerAnnotationRegexPrefix = "~*"
+	// TODO: list all available matchers instead.
```
I read related RFCs, there could be custom HTTP methods, so we should not limit values of methods. removed this TODO.

randmonkey commented:
> That's a lot of good work! +1
> 
> My comments mainly gravitate around my thought that we should be careful when adding such a volume of brand new code by making sure that all of the pieces we add are properly unit tested + we have at least one integration test covering it works with Kong as expected in a sunny day scenario.

I added several unit tests fro translating ingresses to expression routes. For integration tests, I added workflow `integartion-tests/dbless-expression-router` to run integration tests that are valid using expression routes. tests using `TCPIngress`,`UDPIngress`,`TCPRoute`,`HTTPRoute`,`GRPCRoutes` and `knative.Ingress` are skipped. Also, the case testing HTTP redirect is skipped because Kong gateway does not support this when using expression router.

pmalek commented on internal/dataplane/kong_client.go:
```diff
@@ -63,6 +63,10 @@ type KongClient struct {
 	// the newer logic which combines them.
 	enableCombinedServiceRoutes bool
 
+	// enableExpressionRoutes indacates whether the data-plane client will
```
```suggestion
	// enableExpressionRoutes indicates whether the data-plane client will
```

pmalek commented on internal/dataplane/kongstate/route.go:
```diff
@@ -218,18 +232,21 @@ func (r *Route) overrideSNIs(log logrus.FieldLogger, anns map[string]string) {
-	r.overrideHeaders(r.Ingress.Annotations)
-	r.overridePathHandling(log, r.Ingress.Annotations)
+
+	if !r.ExpressionRoutes {
```
Maybe add a comment why do we need to skip those overrides ? It's not immediately obvious to me

randmonkey replied:
added comments here.

pmalek commented on test/integration/grpcroute_test.go:
```diff
@@ -65,6 +65,8 @@ func grpcEchoResponds(ctx context.Context, url, hostname, input string) error {
 }
 
 func TestGRPCRouteEssentials(t *testing.T) {
+	// TODO: implement translator for grpc routes
```
Can we add an issue for this and link it here?

randmonkey replied:
added:#3752

pmalek commented on test/integration/helpers_test.go:
```diff
@@ -416,3 +417,11 @@ func setIngressClassNameWithRetry(ctx context.Context, namespace string, obj run
+func skipTestForExpressionRouter(t *testing.T) {
+	t.Helper()
+	routerFlavor := testenv.KongRouterFlavor()
+	if routerFlavor == "expressions" {
```
Do you mind adding a const for this?

randmonkey replied:
added.

pmalek commented on test/integration/helpers_test.go:
```diff
@@ -416,3 +417,11 @@ func setIngressClassNameWithRetry(ctx context.Context, namespace string, obj run
 	return fmt.Errorf("unsupported GroupVersionKind %v", obj.GetObjectKind())
 }
+
+func skipTestForExpressionRouter(t *testing.T) {
```
I like this but it's not clear to me why those are skipped. Is it because of our test logic is not compatible i.e. it expect different objects than would be produced when expressions flavour router is used?

Do you mind adding a comment about this to this func?

randmonkey replied:
added comments.

pmalek commented on internal/manager/run.go:
```diff
@@ -158,6 +161,15 @@ func Run(ctx context.Context, c *Config, diagnostic util.ConfigDumpDiagnostic, d
 	}
 
+	if enabled, ok := featureGates[featuregates.ExpressionRoutesFeature]; ok && enabled {
+		if routerFlavor == "expressions" {
```
Ditto here about using a `const`

randmonkey replied:
used `const`.

czeslavo commented on internal/dataplane/parser/translators/ingress_atc_test.go:
```diff
@@ -0,0 +1,259 @@
+			expression: `http.path ~ "^/[a-z]+"`,
+		},
+		{
+			name: "empty path",
```
```suggestion
			name: "empty prefix path",
```
I think that we could also add two more cases for empty paths for regex and exact types.

randmonkey replied:
added more tests on paths.

czeslavo commented on internal/dataplane/parser/translators/ingress_atc_test.go:
```diff
@@ -0,0 +1,259 @@
+		{
+			name: "regex match",
+			path: netv1.HTTPIngressPath{
+				Path:     "/~/[a-z]+",
```
How about adding one more case for a regex that already includes the `^` prefix, e.g. `/~/^[a-z]+`?

randmonkey replied:
added.

czeslavo commented on internal/dataplane/parser/translators/ingress_atc_test.go:
```diff
@@ -0,0 +1,259 @@
+}
+
+func TestHeaderMatcherFromHeaders(t *testing.T) {
+	testCases := []struct {
```
Can we please add more test cases? That could be e.g.:
- `Host` is skipped,
- Empty headers,
- Multiple values for a single header,
- Regex match for the header provided.

randmonkey replied:
Added more tests.

czeslavo commented on internal/dataplane/parser/translators/ingress_atc_test.go:
```diff
@@ -0,0 +1,259 @@
+			require.Equal(t, tc.expression, matcher.Expression())
+		})
+	}
+}
```
I think we're still missing UTs for below functions, would you mind adding them?
- `hostMatcherFromIngressHosts`
- `protocolMatcherFromProtocols`
- `methodMatcherFromMethods`
- `sniMatcherFromSNIs`.

randmonkey replied:
Added tests for them.

czeslavo commented on internal/manager/utils/kongconfig/root.go:
```diff
@@ -16,54 +16,97 @@ import (
+		return "", err
+	}
+
+	routerFlavor, ok := rootConfig["router_flavor"].(string)
```
Just to be safe from potential panics, we could split this into two steps (same for `DBModeFromRoot`):
```go
const routerFlavorKey = "router_flavor"
routerFlavor, exists := rootConfig[routerFlavorKey]
if !exists {
  return "", fmt.Errorf("no value in root configuration for %q key", routerFlavorKey)
}

routerFlavorStr, ok := routerFlavor.(string)
if !ok {
  return "", fmt.Errorf("invalid %q type, expected a string, got %T", routerFlavorKey, routerFlavor)
}
```

randmonkey replied:
This will NOT cause panic if `router_flavor` is not present in the map when using `routerFlavor, ok := rootConfig["router_flavor"].(string)`. But it is a good suggestion to make the error clearer.

pmalek commented on internal/manager/featuregates/feature_gates.go:
```diff
@@ -26,6 +26,8 @@ const (
 	// objects like Ingress instead of creating a route per path.
 	CombinedRoutesFeature = "CombinedRoutes"
 
+	ExpressionRoutesFeature = "ExpressionRoutes"
```
doc comment for this? Ideally with a link to docs explaining what this actually is

pmalek commented on internal/dataplane/parser/translators/ingress_atc.go:
```diff
@@ -0,0 +1,260 @@
+	matchers := make([]atc.Matcher, 0, len(headers))
+
+	// To make a stable result from the same annotations, sort the header names first.
+	headerNames := []string{}
```
Preallocate?

randmonkey replied:
done

pmalek commented on internal/dataplane/parser/translators/ingress_atc.go:
```diff
@@ -0,0 +1,260 @@
+	matchers := []atc.Matcher{}
+	for _, protocol := range protocols {
+		if !util.ValidateProtocol(protocol) {
+			fmt.Println("invalid protocol: ", protocol)
```
I assume we don't want to print here but rather handle the error/unwanted state somehow.

If we don't need to handle the error then just going straight to `continue` makes sense, otherwise let's return and `error` and handle it

randmonkey replied:
removed the `Println`.

pmalek reviewed (approved):
:shipit:

czeslavo reviewed (approved):
🎉
